File: frontend/services/tracker.ts
import apiClient from "@/services/api";
import { GlobalResponse } from "@/services/type/responses";

export type ServiceStatus = "pending" | "ongoing" | "completed" | "failed";

export type Tracker = {
  ID?: string;
  ticket_id: string;
  branch_id: string;
  service_status: string;
  admission_date: string;
  product_sn: string;
  customer_name: string;
  phone_number: string;
  email: string;
};

export type GetTrackerPayload = {
  limit: number;
  page: number;
  service_status?: ServiceStatus;
};

export const createTracker = async (payload: Tracker) => {
  return apiClient.post<GlobalResponse<any>>(`/api/tracker/`, { payload })
}

export const getTrackers = async (data: GetTrackerPayload) => {
  return apiClient.get<GlobalResponse<Tracker[]>>(`/api/tracker/`, {
    params: {
      ...data,
    },
  });
};

export const updateTrackerStatus = async (
  id: string,
  status: ServiceStatus,
) => {
  return apiClient.put<GlobalResponse<null>>(`/api/tracker/${id}/status`, {
    service_status: status,
  });
};

File: backend-go/models/tracker.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type ServiceStatus string

const (
	ServicePending   ServiceStatus = "pending"
	ServiceOngoing   ServiceStatus = "ongoing"
	ServiceCompleted ServiceStatus = "completed"
	ServiceFailed    ServiceStatus = "failed"
)

type Tracker struct {
	gorm.Model
	TicketID      string        `gorm:"uniqueIndex;not null" json:"ticket_id"`
	ServiceStatus ServiceStatus `gorm:"type:varchar(20);check:service_status IN ('pending','ongoing','completed','failed');default:'pending'" json:"service_status"`
	BranchID      string        `json:"branch_id" gorm:"not null"`
	AdmissionDate time.Time     `gorm:"not null" json:"admission_date"`
	Complains     string        `json:"complains"`
	ProductSN     string        `json:"product_sn"`
	CustomerName  string        `json:"customer_name"`
	PhoneNumber   string        `json:"phone_number"`
	Email         string        `json:"email"`
}
